Sources/_MatchingEngine/Utility/MissingUnicode.swift

```diff
@@ -174,21 +174,21 @@ extension Unicode {
     case zanabazarSquare = "Zanabazar_Square"
   }
 
-  public enum POSIXCharacterSet: String, Hashable {
+  /// POSIX character properties not already covered by general categories or
+  /// binary properties.
+  public enum POSIXProperty: String, Hashable {
     case alnum = "alnum"
-    case alpha = "alpha"
-    case ascii = "ascii"
     case blank = "blank"
-    case cntrl = "cntrl"
-    case digit = "digit"
     case graph = "graph"
-    case lower = "lower"
     case print = "print"
-    case punct = "punct"
-    case space = "space"
-    case upper = "upper"
     case word = "word"
     case xdigit = "xdigit"
+    // As per http://www.unicode.org/reports/tr18/#Compatibility_Properties,
+    // [:alpha:], [:lower:], [:upper:], and [:space:] are covered by binary
+    // properties. [:punct:], [:digit:], and [:cntrl:] are covered by general
+    // categories. [:ascii:] is covered by CharacterProperty.Kind.ascii.
+    // These may have different semantics depending on matching mode, but that
+    // should be left up to the matching engine.
```

Comment:
But, people can still type them in and we'll still parse them right? Wouldn't we want to preserve how they specified it, especially since the interpretation varies based on semantic mode?

Reply:
Yes we still parse them, and we preserve whether they were written using `[:...:]` or `\p{...}` syntax, which could be used to vary behavior based on matching mode. It's just that by default they will be classified as general categories or binary properties, as Unicode defines aliases for them that cover the POSIX spellings. We could prefer to classify them as POSIX, but IMO it's nicer to have a complete Unicode syntax and add in the bits that it doesn't cover, what do you think? It's also complicated by the fact that the Unicode definitions have aliases for e.g `punct`, `punctuation`, and `p`. It may be that we need to preserve which alias was used, as e.g should only the first get the POSIX behavior?

Reply:
The latest Nate and I talked, we were thinking that `[:...:]` character classes would be restricted to their traditional ASCII definition regardless of semantic mode, `\p{...}` character classes would be restricted to their traditional Unicode definition regardless of semantic mode, and `\w` style character classes would be more flexible binding to ASCII, Unicode, or grapheme cluster definitions depending on semantic mode.

Reply:
Do we have any rationale why that particular syntax would deviate from _both_ the normal mechanisms for reasoning about semantic levels _and_ Unicode's explicit recommendations in [UTS-18 Annex C](https://www.unicode.org/reports/tr18/#Compatibility_Properties)?

Reply:
IMO the shortest spelling feels the most underspecified. I think we can defend it more at length in the proposal. FWIW It seems like this decision is precedented in Perl: https://perldoc.perl.org/perlre#/a-(and-/aa)

> **/a (and /aa)**
> 
> This modifier stands for ASCII-restrict (or ASCII-safe). This modifier may be doubled-up to increase its effect.
> 
> When it appears singly, it causes the sequences \d, \s, \w, and the Posix character classes to match only in the ASCII range.

Reply:
By default, perl will follow the model we're outlining here. If you specify ASCII mode in that way (or posix mode in some engines), then the POSIX-syntactic classes will be restricted to ASCII. The remaining question is whether `\p` should as well, which I don't think is a question we'll be answering unilaterally, and there are contextual options controlling this:

https://github.com/kkos/oniguruma/blob/master/doc/RE#L270

```
  (?imxWDSPy-imxWDSP:subexp)  option on/off for subexp

                           i: ignore case
                           m: multi-line (dot (.) also matches newline)
                           x: extended form
                           W: ASCII only word (\w, \p{Word}, [[:word:]])
                              ASCII only word bound (\b)
                           D: ASCII only digit (\d, \p{Digit}, [[:digit:]])
                           S: ASCII only space (\s, \p{Space}, [[:space:]])
                           P: ASCII only POSIX properties (includes W,D,S)
                              (alnum, alpha, blank, cntrl, digit, graph,
                               lower, print, punct, space, upper, xdigit, word)

                           y{?}: Text Segment mode
                              This option changes the meaning of \X, \y, \Y.
                              Currently, this option is supported in Unicode only.

                              y{g}: Extended Grapheme Cluster mode (default)
                              y{w}: Word mode
                              See [Unicode Standard Annex #29]

```

Either way, this affects the AST representation to the extent we record these options and present API for reasoning. I don't think we have a good enough feel at this moment to decide whether it's a different syntactic node or not. Most of the AST API relevant to users is along a more semantic (and perhaps parameterized) level instead of derived via a parse tree.